File: 2021/05/17/main.py
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        self.mp = {}
        self.ans = 0
        for w in words:
            self.mp[w] = 1
        words.sort(key=len)
        for w in words:
            for i in range(len(w)):
                p = str(w[:i] + w[i+1:])
                if p in self.mp and self.mp[p] + 1 > self.mp[w]:
                    self.mp[w] = self.mp[p] + 1
            if self.mp[w] > self.ans:
                self.ans = self.mp[w]
        return self.ans
        

File: 2021/05/17/main.go
package main

import "sort"

func longestStrChain(words []string) int {
	mp := map[string]int{}
	ans := 0
	for _, w := range words {
		mp[w] = 1
	}
	sort.Slice(words, func(i, j int) bool {
		return len(words[i]) < len(words[j])
	})

	for _, w := range words {
		for i := 0; i < len(w); i++ {
			p := w[:i] + w[i+1:]
			_, exists := mp[p]
			if exists && mp[p]+1 > mp[w] {
				mp[w] = mp[p] + 1
			}
		}
		if mp[w] > ans {
			ans = mp[w]
		}
	}
	return ans
}
